ipq40xx: add support for EnGenius EMR3500

Yen-Ting-Shen opened the pull request:
SOC:     IPQ4018 / QCA Dakota
    CPU:     Quad-Core ARMv7 Processor rev 5 (v7l) Cortex-A7
    DRAM:    256 MiB
    NOR:     32 MiB
    ETH:     Qualcomm Atheros QCA8072 (2 ports)
    USB:     1 x 2.0 (Host controller in the SoC)
    WLAN1:   Qualcomm Atheros QCA4018 2.4GHz 802.11bgn 2:2x2
    WLAN2:   Qualcomm Atheros QCA4018 5GHz 802.11a/n/ac 2:2x2
    INPUT:   RESET Button
    LEDS:    White, Blue, Red, Orange
    Buttons: 1 x Reset button

    Flash instruction:

    From EnGenius firmware to OpenWrt firmware:

    In Firmware Upgrade page, upgrade your openwrt-ipq40xx-generic-engenius_emr3500-squashfs-factory.bin directly.

    From OpenWrt firmware to EnGenius firmware:

    1. Setup a TFTP server on your computer and configure static IP to 192.168.99.8
       Put the EnGenius firmware in the TFTP server directory on your computer.
    2. Power up EMR3500. Press 4 and then press any key to enter u-boot.
    3. Download EnGenius firmware
       (IPQ40xx) # tftpboot 0x84000000 openwrt-ipq40xx-emr3500-nor-fw-s.img
    4. Flash the firmware
       (IPQ40xx) # imgaddr=0x84000000 && source 0x84000000:script
    5. Reboot
       (IPQ40xx) # reset

Signed-off-by: Yen-Ting-Shen <frank.shen@senao.com>

Thanks for your contribution to OpenWrt!

To help keep the codebase consistent and readable,
and to help people review your contribution,
we ask you to follow the rules you find in the wiki at this link
https://openwrt.org/submitting-patches

Please remove this message before posting the pull request.

Yen-Ting-Shen commented on target/linux/ipq40xx/base-files/etc/hotplug.d/firmware/11-ath10k-caldata:
```diff
@@ -164,6 +164,10 @@ case "$FIRMWARE" in
 		;;
+	engenius,emr3500)
+        ath10kcal_extract "0:ART" 0x1000 0x2f20
+        ath10kcal_patch_mac_crc $(cat /sys/class/net/eth0/address)
```
I've updated the file to correct the indent.
There's no MAC address for this WIFI in 0:APPSBLENV.

Yen-Ting-Shen replied:
The stock firmware uses LAN MAC address as 2.4G wifi MAC address.

Yen-Ting-Shen commented on target/linux/ipq40xx/base-files/etc/hotplug.d/firmware/11-ath10k-caldata:
```diff
@@ -225,6 +229,10 @@ case "$FIRMWARE" in
 		;;
+	engenius,emr3500)
+		ath10kcal_extract "0:ART" 0x5000 0x2f20
+		ath10kcal_patch_mac_crc $(macaddr_add "$(cat /sys/class/net/eth0/address)" 1)
```
It's not stored in 0:APPSBLENV.

Yen-Ting-Shen replied:
The stock firmware uses (LAN MAC address + 1) as 5G wifi MAC address.

adrianschmutzler replied:
Okay. Is eth0 the same as one of $(mtd_get_mac_ascii 0:APPSBLENV wanaddr) or $(mtd_get_mac_ascii 0:APPSBLENV ethaddr)?
In that case, I would prefer patching WiFi MAC based on those instead of $(cat /sys/class/net/eth0/address) ...

Yen-Ting-Shen replied:
OK. done.
I also use this way to patch MAC address for 2.4G above.

adrianschmutzler commented:
Consider adding label_mac as discussed in your other commit...

Yen-Ting-Shen commented:
> Consider adding label_mac as discussed in your other commit...

Added.

Yen-Ting-Shen commented:
Hi,
Does there anything still needs changing? Can this PR go to the next step?

adrianschmutzler commented:
Please rebase onto recent master.

Yen-Ting-Shen commented:
This PR has been rebased.

adrianschmutzler commented:
You have an indent in commit message, please remove that one.

Yen-Ting-Shen committed:
ipq40xx: add support for EnGenius EMR3500

SOC:     IPQ4018 / QCA Dakota
CPU:     Quad-Core ARMv7 Processor rev 5 (v7l) Cortex-A7
DRAM:    256 MiB
NOR:     32 MiB
ETH:     Qualcomm Atheros QCA8072 (2 ports)
USB:     1 x 2.0 (Host controller in the SoC)
WLAN1:   Qualcomm Atheros QCA4018 2.4GHz 802.11bgn 2:2x2
WLAN2:   Qualcomm Atheros QCA4018 5GHz 802.11a/n/ac 2:2x2
INPUT:   RESET Button
LEDS:    White, Blue, Red, Orange

Flash instruction:

From EnGenius firmware to OpenWrt firmware:

In Firmware Upgrade page, upgrade your openwrt-ipq40xx-generic-engenius_emr3500-squashfs-factory.bin directly.

From OpenWrt firmware to EnGenius firmware:

1. Setup a TFTP server on your computer and configure static IP to 192.168.99.8
   Put the EnGenius firmware in the TFTP server directory on your computer.
2. Power up EMR3500. Press 4 and then press any key to enter u-boot.
3. Download EnGenius firmware
   (IPQ40xx) # tftpboot 0x84000000 openwrt-ipq40xx-emr3500-nor-fw-s.img
4. Flash the firmware
   (IPQ40xx) # imgaddr=0x84000000 && source 0x84000000:script
5. Reboot
   (IPQ40xx) # reset

Signed-off-by: Yen-Ting-Shen <frank.shen@senao.com>

Yen-Ting-Shen commented:
removed!

chunkeey commented:
@blogic, since you are currently assigned on this, can I merge some of these?

chunkeey commented:
@blogic, ping!

Yen-Ting-Shen committed:
ipq40xx: add support for EnGenius EMR3500 - update

chunkeey commented:
@blogic, ping! ping! ping!

chunkeey commented:
@blogic last ping!

ynezz commented:
@chunkeey I've pinged blogic on IRC, unassigned, thanks for taking care!

chunkeey commented:
@ynezz thanks. I've updated the patch and just merged it...
